### PR TITLE
executor: fix wrong behavior when `ENUM` column meet Aggregation

### PR DESCRIPTION
**Pull request by pingyu**
### What problem does this PR solve? 
Fix wrong behavior when `ENUM` column meet Aggregation as #13027

### What is changed and how it works?
Add `firstRow4Enum`, the `first_row` executor for `ENUM`

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
```sql
mysql> create table t(a enum('a', 'b'));
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values('a'), ('b');
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> select cast(a as signed) from (select a from t group by a) t;
+-------------------+
| cast(a as signed) |
+-------------------+
|                 1 |
|                 2 |
+-------------------+
2 rows in set (0.00 sec)

mysql> 
```

### CODE REVIEW

**Comment by winoros** [comment]
@pingyu From planner's perspective, other function is okay, will not cause such obvious bug.

**Comment by SunRunAway** [comment]
> @pingyu Yes, this is what i want!
> 
> @SunRunAway From planner's perspective, other function is okay, will not cause such obvious bug.
> 
> Only the `FirstRow` function is eliminated in the outputing columns of aggregation(i.e. For other funcs, we'll create a new column to present the aggregate function, but we use the original column instead of a new column to present `FirstRow` function).
> Thus if one aggregation can be eliminated since its group by items are not null and unique, the wrong type inferring would lead to wrong optimizing result.

Thanks.
Could you also add implementation for `TypeSet`? @pingyu

**Comment by pingyu** [comment]
Sure, with pleasure~ @SunRunAway

**Comment by pingyu** [comment]
Implementation for `TypeSet` added. PTAL, thanks~ @SunRunAway 

```
mysql> create table t1(f set('a', 'b', 'c'));
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t1 values('a'), ('b'), ('a,b,c');
Query OK, 3 rows affected (0.01 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> select f, cast(f as signed) from (select f from t1 group by f) t;
+-------+-------------------+
| f     | cast(f as signed) |
+-------+-------------------+
| a,b,c |                 7 |
| b     |                 2 |
| a     |                 1 |
+-------+-------------------+
3 rows in set (0.00 sec)

mysql> 
```

File: expression/aggregation/base_func.go [hunk 1]
```diff
@@ -188,7 +188,8 @@ func (a *baseFuncDesc) typeInfer4MaxMin(ctx sessionctx.Context) {
 		a.RetTp = a.Args[0].GetType().Clone()
 		a.RetTp.Flag &^= mysql.NotNullFlag
 	}
-	if a.RetTp.Tp == mysql.TypeEnum || a.RetTp.Tp == mysql.TypeSet {
+	// TODO: fix other aggFuncs for TypeEnum & TypeSet
```
Comment (SunRunAway) [inline expression/aggregation/base_func.go]: According to https://github.com/pingcap/tidb/pull/14035#issuecomment-566505617, we do not need this `TODO`.
And we can add a annotation to describe why we need this judgment.

Comment (pingyu) [reply expression/aggregation/base_func.go]: Function `max()` for `enum` in aggregation is incorrect either, as represented in #13961 .
So I think `max`, `min`, `first_value`, `last_value`, `nth_value` for `enum` and `set` need such a fix too.
Maybe I can try to do it using `go:generate`.

**Comment by SunRunAway** [comment]
@pingyu Could you help cherry-pick it to release-2.1 branch ?

**Pull request merged**